### PR TITLE
ENT4379: Route to default_idp if no tpa_hint is added in query_param

### PR DESCRIPTION
**Pull request by zamanafzal**
**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: enterprise/models.py [hunk 1]
```diff
@@ -453,6 +453,8 @@ def get_tpa_hint(self, tpa_hint_param=None):
         # if there is only one identity provider linked with the customer and tpa_hint_param was not provider.
         if self.has_single_idp and not tpa_hint_param:
             return self.identity_providers.first().provider_id
+        if self.has_multiple_idps and not tpa_hint_param:
+            return self.default_provider_idp.provider_id
```
Comment (HammadAhmadWaqas) [inline enterprise/models.py:457]: For the safe side we should check if the `self.default_provider_idp` exists before excessing its `provider_id`.

File: enterprise/models.py [hunk 2]
```diff
@@ -677,8 +679,10 @@ def notify_enrolled_learners(self, catalog_api_user, course_id, users, admin_enr
         course_path = '/courses/{course_id}/course'.format(course_id=course_id)
         params = {}
         # add tap_hint if there is only one IdP attached with enterprise_customer
-        if self.identity_providers.count() == 1:
+        if self.has_single_idp:
             params = {'tpa_hint': self.identity_providers.first().provider_id}
+        elif self.has_multiple_idps:
+            params = {'tpa_hint': self.default_provider_idp.provider_id}
```
Comment (HammadAhmadWaqas) [inline enterprise/models.py:685]: same concern as above.

File: tests/test_models.py [hunk 3]
```diff
@@ -234,6 +234,40 @@ def test_identity_providers(self):
...
+            default_provider=True
+        )
+        ent_idp_two = factories.EnterpriseCustomerIdentityProviderFactory(enterprise_customer=customer)
+        assert customer.get_tpa_hint() == ent_idp_one.provider_id
+        ent_idp_one.default_provider = False
+        ent_idp_one.save()
+        ent_idp_two.default_provider = True
+        ent_idp_two.save()
+        assert customer.get_tpa_hint() == ent_idp_two.provider_id
+
```
Comment (HammadAhmadWaqas) [inline tests/test_models.py:270]: Can we add a test if there are multiple attached but not default IDP exists.

**Review by HammadAhmadWaqas — CHANGES_REQUESTED** [review]

File: enterprise/models.py (hunk 1 above)
Comment (zamanafzal) [reply enterprise/models.py:457]: If an enterprise wants to link more than one IDP then it has to made default one of them. Otherwise, it can not link it.

Comment (zamanafzal) [reply enterprise/models.py:457]: Screenshot 2021-06-22 at 2 31 45 PM

Comment (HammadAhmadWaqas) [reply enterprise/models.py:457]: @zamanafzal You are 100% right, we can't link multiple IDPs without defining default from the admin, But this restriction is only in admin view. But this constraint is not on the Model level. In future, there might be scenarios when we link multiple IDPs with default so I would recommend better checking if default IDP is there before using it.

Comment (zamanafzal) [reply enterprise/models.py]: I think this behaviour is not expected if we look at the default IDP ticket. But still for the safe side, I have added this check and test case accordingly.

**Review by HammadAhmadWaqas — APPROVED** [review]

File: enterprise/models.py [hunk 4]
```diff
@@ -677,8 +679,11 @@ def notify_enrolled_learners(self, catalog_api_user, course_id, users, admin_enr
         course_path = '/courses/{course_id}/course'.format(course_id=course_id)
         params = {}
         # add tap_hint if there is only one IdP attached with enterprise_customer
-        if self.identity_providers.count() == 1:
+        if self.has_single_idp:
             params = {'tpa_hint': self.identity_providers.first().provider_id}
+
+        elif self.has_multiple_idps and self.default_provider_idp:
```
Comment (muhammad-ammar) [inline enterprise/models.py:685]: What if there are `multiple idps` but there is no `default provider idp`?

**Review by muhammad-ammar — APPROVED** [review]

**Review by saleem-latif — APPROVED** [review]

File: enterprise/models.py (hunk 4 above)
Comment (zamanafzal) [reply enterprise/models.py:685]: This is discussed at the first comment in the PR.
https://github.com/edx/edx-enterprise/pull/1276#discussion_r656038394

**Pull request merged**